source/agora/common/Hash.d

```diff
@@ -0,0 +1,159 @@
+    a `hash` function to support hashing.
+    This limitation might be lifted in the future, but a few things
+    need to be taken into account:
+    - Alignment / packing: We can't have unaligned / unpacked structures
+      as it would create malleability issues
+    - Non-public members: We can't deal with anything non-public (we can use
+      `.tupleof` but it's impossible to tell the usage of the member,
+      e.g. it could be a cache of the hash)
+    - Indirections / references types
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.Hash;
+
+import agora.common.Data;
+
+import libsodium;
+
+
+///
+nothrow @nogc @safe unittest
+{
+    static assert(crypto_generichash_BYTES == Hash.sizeof);
+    auto r1 = hashFull("abc");
+    assert(r1 == hashFull("abc"));
+
+    static struct Struct
+    {
+        ulong foo;
+        string bar;
+
+        void computeHash (scope HashDg dg) const nothrow @safe @nogc
+        {
+            // We can hash in any order we want
+            hashPart(this.bar, dg);
+            hashPart(this.foo, dg);
+        }
+    }
+
+    Struct st = Struct(42, "42");
+    auto r2 = hashFull(st);
+    assert(hashFull(Struct(42, "42")) == r2);
+
+    auto r3 = hashFull(r1);
+    auto r3_exp = Hash(`0xf3f428948cc7a089291522483a416324a1457d6c6b55aee0ea830bedaa4c67f6`);
+    assert(r3 == r3_exp);
+
+    // TODO: Add a check with some well-known outputs
+}
+
+
+/// Type of delegate passed to `hash` function when there's a state
+public alias HashDg = void delegate(scope const(ubyte)[]) /*pure*/ nothrow @safe @nogc;
+
+
+/*******************************************************************************
+
+    Hash a given data structure using BLAKE2b into `state`
+
+    Note that there is no overload for signed types, as only the binary
+    representation matters for hashing.
+
+    Params:
+      T = Type of struct to hash
+      record = Instance of `T` to hash
+      state  = State delegate, when this struct is nested in another.
+
+    Returns:
+      The `Hash` representing this instance
+
+*******************************************************************************/
+
+public Hash hashFull (T) (scope const auto ref T record)
+    nothrow @nogc @trusted
+{
+    Hash hash = void;
+    crypto_generichash_state state;
+    crypto_generichash_init(&state, null, 0, Hash.sizeof);
+    scope HashDg dg = (scope const(ubyte)[] data) @trusted
+                     => cast(void)crypto_generichash_update(&state, data.ptr, data.length);
+    hashPart(record, dg);
+    crypto_generichash_final(&state, hash[].ptr, Hash.sizeof);
+    return hash;
+}
+
+/// Ditto
+public void hashPart (T) (scope const auto ref T record, scope HashDg state)
```

Comment:
Shouldn't these be private? Or not?

Reply:
Ah no, I get it now. It's called bye the struct's implementation of `computeHash`.